enhancements/microshift/audit-log-configuration-options.md

```diff
@@ -0,0 +1,249 @@
+
+```yaml
+apiServer:
+  auditLog:
+    maxFileSize: STRING
+    maxFiles: INT
+    maxFileAge: STRING
+```
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+- N/A
+
+#### Standalone Clusters
+
+- N/A
+
+#### Single-node Deployments or MicroShift
+
+- This EP is specific to MicroShift only.
+
+### Implementation Details/Notes/Constraints
+
+**Passing User Provided Options to the API Server**
+MicroShift embeds a default API server config, written to disk during boot at `/var/lib/microshift/resources/kube-apiserver-audit-policies/default.yaml`. The file is structured as a `KubeAPIServerConfig` custom resource, provided by the `kubecontrolplane.config.openshift.io/v1` API group. During boot, the kube-apiserver service-manager collects kube-apiserver options from hardcoded defaults and merges these into the kube-apiserver config data structure.  The override values are iteratively patch onto the embedded default config.  The order of precedence is from first to last, such that field paths appearing later in the queue override data fields from previous iterations.  The logic which handles the overriding is [here](https://github.com/openshift/microshift/blob/ba79f993070111ae3b6e35c3abb3141e95901874/pkg/controllers/kube-apiserver.go#L226-L250). Thus, to pass user-defined audit configuration options to the API server, MicroShift will append the user-specified options to `overrides` array. This minimizes the complexity of the option-passing and allows for MicroShift defaults to continue to be specified as they are now.  This also provides an elegant fallback behavior for user-facing fields. If a user does not specify a value for a field, the default value will be used, just as it is now.  Should a user remove a field they previously set, then at next restart, the field will no longer be overridden, thus passing the default. All audit logging configuration fields must be optional. If not specified, the existing default audit logging policy will be used.
+
+**Audit Log Rotation**
+
+The kube-apiserver provides three CLI flags to dictate log rotation and retention policies which may be thinly exposed to the user via the MicroShift config API. These options will be set dynamically, depending on user configuration in the MicroShift config. Logic already exists to set these flags with hardcoded values, which will need to be updated to allow for dynamic arguments.  These options are:
+
+* `--audit-log-maxage` defines the maximum number of days to retain old audit log files.  MicroShift defaults to 10 days.
+* `--audit-log-maxbackup` defines the maximum number of audit log files to retain. MicroShift defaults to 10 files.
+* `--audit-log-maxsize` defines the maximum size in megabytes the live log file before it is rotated.  Defaults to 200Mb.
+
+Thus, the default maximum storage consumption of audit logs will be 2000Mb, provided all files are younger than 10 days.
```

Comment:
Should we decrease the default to 1GB instead?

Reply:
Eh... the default values here are set by the kube-apiserver. Unless halving the default size is a customer ask, I'm not in favor of overriding it arbitrarily.

Reply:
OK, the current ER may not be related to this, but we might want to keep an eye on the 2GB limit which is a bit on the high side for edge devices. We can follow-up separately as a bug if this turns out to be a problem.
Please, resolve this comment.